ant-networking/src/event/identify.rs

```diff
@@ -6,187 +6,236 @@
-                if !has_relayed && !is_bootstrap_peer {
-                    if let Some(relay_manager) = self.relay_manager.as_mut() {
-                        debug!(
-                            "Adding candidate relay server {peer_id:?}, it's not a bootstrap node"
-                        );
-                        relay_manager.add_potential_candidates(&peer_id, &addrs, &info.protocols);
-                    }
-                }
+                    .any(|entry| entry.node.key.preimage() == &peer_id);
+
+                // // If the bucket contains any of a bootstrap node,
+                // // consider the bucket is not full and dial back
+                // // so that the bootstrap nodes can be replaced.
+                // if is_bucket_full {
+                //     if let Some(peers) = self.bootstrap_peers.get(&ilog2) {
+                //         if kbucket.iter().any(|entry| {
+                //             peers.contains(entry.node.key.preimage())
+                //         }) {
+                //             is_bucket_full = false;
+                //         }
+                //     }
+                // }
+
+                (is_bucket_full, already_present_in_rt, ilog2)
+            } else {
+                return;
+            };
+
+        // If the peer is part already of the RT, try updating the addresses based on the new push info.
+        // We don't have to dial it back.
+        if already_present_in_rt {
+            self.update_pre_existing_peer(peer_id, addrs.clone());
+            return;
+        }
+
+        // When received an identify from un-dialed peer, try to dial it
+        // The dial shall trigger the same identify to be sent again and confirm
+        // peer is external accessible, hence safe to be added into RT.
+        if !self.local && !has_dialed {
+            // Only need to dial back for not fulfilled kbucket
+            if kbucket_full {
+                debug!("received identify for a full bucket {ilog2:?}, not dialing {peer_id:?} on {addrs:?}");
+                return;
+            } else if already_present_in_rt {
+                debug!("received identify for {peer_id:?} that is already part of the RT. Not dialing {peer_id:?} on {addrs:?}");
+                return;
+            }
+
+            info!(%peer_id, ?addrs, "received identify info from undialed peer for not full kbucket {ilog2:?}, dial back to confirm external accessible");
+            if let Err(err) = self.swarm.dial(
+                DialOpts::peer_id(peer_id)
+                    .condition(PeerCondition::NotDialing)
+                    .addresses(addrs.iter().cloned().collect())
+                    .build(),
+            ) {
+                warn!(%peer_id, ?addrs, "dialing error: {err:?}");
+            }
+
+            return;
+        }
+
+        // If we are not local, we care only for peers that we dialed and thus are reachable.
```

Comment:
comment doesn't match the condition check (i.e. `self.local || has_dialed` ) ?
also `self.local || has_dialed` is just the `else` of `!self.local && !has_dialed` , why make it separate `if clause` ?

Reply:
I haven't modified anything there. The comment has to be updated there. I think we should add the peers directly to the RT if local.

Reply:
Also added an else there.